### PR TITLE
Make IndexAnalyzers an interface

### PR DESCRIPTION
**Pull request by romseygeek**
IndexAnalyzers is currently always a concrete class wrapping several
Maps of NamedAnalyzers.  This means that whenever it is used it needs
to instantiate all of its component analyzers, making testing much heavier
than it needs to be.  It also means that things like overriding analysis for
legacy indexes is pushed into mapper parameters, rather than being 
handled in a single place.

This commit makes IndexAnalyzers into an interface, with an anonymous
concrete implementation that handles reloading and closing for index
shards.

### CODE REVIEW

**Comment by romseygeek** [comment]
This opens up a number of nice future improvements:
* legacy index analysis handling can be moved to a single IndexAnalyzers wrapper, rather than being scattered across the various field mappers that have to handle it
* We can avoid instantiating Analyzers entirely during mapping validation, instead using a 'validating' IndexAnalyzers which checks that analyzers are correctly defined in index settings but avoiding having to actually build anything.
* We can move Analyzer handling from mapper parsing to mapper building, making it much easier to handle analysis in _all fields.

**Comment by romseygeek** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java [hunk 1]
```diff
@@ -649,6 +650,9 @@ public static IndexAnalyzers build(
             throw new IllegalArgumentException("no default analyzer configured");
         }
         defaultAnalyzer.checkAllowedInMode(AnalysisMode.ALL);
+        if (Objects.equals(defaultAnalyzer.name(), DEFAULT_ANALYZER_NAME) == false) {
+            throw new IllegalStateException("default analyzer must have the name [default] but was: [" + defaultAnalyzer.name() + "]");
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java:654]: Can we add a new check for this exception? The existing one was removed with "testAnalyzerMapChecks" I think and I couldn't find a new one.

File: server/src/main/java/org/elasticsearch/index/analysis/IndexAnalyzers.java [hunk 2]
```diff
@@ -27,83 +27,142 @@
...
+                return switch (type) {
+                    case ANALYZER -> analyzers.get(name);
+                    case NORMALIZER -> normalizers.get(name);
+                    case WHITESPACE -> whitespaceNormalizers.get(name);
+                };
+            }
+
+            @Override
+            public void close() throws IOException {
+                if (closeable) {
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/analysis/IndexAnalyzers.java:130]: Could you explain quickly (and maybe add javadocs to the interface) when this flag should be set and when not? I see it set to true in MapperServiceFactory but to "false" in the several "IndexAnalyzer.of" methods. Some hin on how to use it in the docs would be great.

File: server/src/test/java/org/elasticsearch/index/analysis/IndexAnalyzersTests.java [hunk 3]
```diff
@@ -21,38 +21,13 @@
...
-     * test the checks in the constructor
-     */
-    public void testAnalyzerMapChecks() {
-        Map<String, NamedAnalyzer> analyzers = new HashMap<>();
-        {
-            NullPointerException ex = expectThrows(
-                NullPointerException.class,
-                () -> new IndexAnalyzers(analyzers, Collections.emptyMap(), Collections.emptyMap())
-            );
-            assertEquals("the default analyzer must be set", ex.getMessage());
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/index/analysis/IndexAnalyzersTests.java:34]: I might have missed this, but has this test moved somewhere?

File: server/src/test/java/org/elasticsearch/index/mapper/DynamicTemplatesTests.java [hunk 4]
```diff
@@ -508,6 +513,11 @@ public void testIllegalDynamicTemplateUnknownAttributeRuntime() throws Exception
         assertEquals("unknown parameter [foo] on runtime field [__dynamic__my_template] of type [keyword]", e.getRootCause().getMessage());
     }
 
+    @Override
+    protected IndexAnalyzers createIndexAnalyzers(IndexSettings indexSettings) {
+        return IndexAnalyzers.of(Map.of("default", new NamedAnalyzer("default", AnalyzerScope.INDEX, new StandardAnalyzer())));
+    }
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/index/mapper/DynamicTemplatesTests.java:519]: Why wasn't this override necessary before?

**Review by cbuescher — CHANGES_REQUESTED** [review]
I like this change, left a few minor questions.

File: server/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java (hunk 1 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java:654]: Looking more carefully at how these are constructed, I don't think it's actually possible for the analyzer registered under `default` to *not* be called `default`.  I'll add an assertion, but will keep the exception in place for paranoia's sake.

File: server/src/main/java/org/elasticsearch/index/analysis/IndexAnalyzers.java (hunk 2 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/analysis/IndexAnalyzers.java:130]: This is actually a leftover of a previous go at this, I'll remove it.

File: server/src/test/java/org/elasticsearch/index/analysis/IndexAnalyzersTests.java (hunk 3 above)
Comment (romseygeek) [reply server/src/test/java/org/elasticsearch/index/analysis/IndexAnalyzersTests.java:34]: It's already tested implicitly in AnalysisRegistryTests, but I'll add something more explicit.  Basically, if nothing is explicitly configured as `default` then the AnalysisRegistry will always add one when building an `IndexAnalyzers`, so this was testing for something that couldn't happen in production code.

**Review by cbuescher — APPROVED** [review]
Thanks for the changes, LGTM from my side now.

**Pull request merged**